platform/engineering/design-docs/2021-08-01-autonomous-deployments.md

```diff
@@ -0,0 +1,220 @@
+    - [Alternatives](#alternatives)
+    - [Post MVP](#post-mvp)
+    - [Questions](#questions)
+    - [Revision History](#revision-history)
+
+## Overview
+
+The document's purpose is to outline the minimum viable product (MVP) outcome to create autonomous deployment within the vets-website repository. The end of this document contains a post-MVP plan as well.
+
+### Objective
+
+VFS teams should be able to safely and reliably deploy their frontend applications to any environment at any time. This should happen without waiting for a scheduled deployment or manual intervention by Platform or other VFS teams. Deployments should also not cause significant risk to the build or to the site.
+
+While adding the ability to autonomously deploy, VA should still maintain the benefits of shared components (i.e. VFS teams are still using shared patterns/components for standardization and quality purposes, and we don’t see wide variance from one VA.gov experience to another).
+
+VFS teams should also be able to easily revert their change on their own, as needed.
+
+### Desired Outcomes
+
+- Increasing the production deployment rate
+- Increasing customer satisfaction
+- Decreasing support volume
+- Decreasing the average time from kickoff to production, by initiative complexity
+- Decreasing build failures
+
+### Background
+
+The current deployment model is to merge changes into the `master` branch to get changes deployed to production, staging, and dev servers. Commits to staging and dev happen with relative frequency, but deployments to production are done once a day for all teams/applications through a central platform-level action. If errors occur at the application level, teams need to request an [out-of-band deployment](https://depo-platform-documentation.scrollhelp.site/support/Deployment-Policies.1279787009.html#DeploymentPolicies-Requestingout-of-banddeploys), which needs to be approved by relevant stakeholders. If the error is not critical, the fix needs to wait a full day until the next scheduled deployment.
+
+This model has increased platform support volume, deployment failures, and decreased customer satisfaction while limiting the deployments to only a handful a week. It has also increased support volume for on call teams when out-of-band deployments are needed outside business hours.
+
+Our goal with this initiative is to allow teams to freely deploy their own code, breaking free of the current model.
+
+**We are writing this with the assumption that Github Actions work will continue to completion.**
+
+This phase of the initiative does not include
+- Programmatic isolation of the applications
+- Physical isolation of the applications
+
+### High Level Design
+
+Deployment environments will be remapped from `master` as a central release location to related environment branches. `master` will deploy to production (va.gov), `staging` will deploy to staging (staging.va.gov), and `dev` will deploy to dev (dev.va.gov).
+
```

Comment:
Question:
Is it OK to let people to deploy to any environment in any order?
For instance, is it OK to have code deployed to prod that hasn't been deployed to staging?
My inclination is to say no, simply because I don't quite understand the implications, and later we can change this decision if we see a need.
If we agree on the above, is it just a policy or do we enforce it via code?